rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py

```diff
@@ -69,241 +85,409 @@ def required_components(cls) -> List[Type[Component]]:
-
-        # sort items to ensure same order every time (for tests)
-        ordered_feature_vocabulary = OrderedDict(sorted(feature_vocabulary.items()))
-
-        return ordered_feature_vocabulary
-
-    def _create_sparse_features(self, message: Message) -> None:
-        """Convert incoming messages into sparse features."""
-        import scipy.sparse
-
-        tokens = message.get(TOKENS_NAMES[TEXT])
-        # this check is required because there might be training data examples without
-        # TEXT, e.g., `Message("", {action_name: "action_listen"})`
-        if tokens:
-
-            sentence_features = self._tokens_to_features(tokens)
-            one_hot_seq_feature_vector = self._features_to_one_hot(sentence_features)
-
-            sequence_features = scipy.sparse.coo_matrix(one_hot_seq_feature_vector)
-
-            final_sequence_features = Features(
-                sequence_features,
-                FEATURE_TYPE_SEQUENCE,
-                TEXT,
-                self.component_config[FEATURIZER_CLASS_ALIAS],
-            )
-            message.add_features(final_sequence_features)
-
-    def _tokens_to_features(self, tokens: List[Token]) -> List[Dict[Text, Any]]:
-        """Convert words into discrete features."""
-
-        configured_features = self.component_config["features"]
+                for position_and_feature_name, feature_value in token_features.items():
+                    feature_vocabulary.setdefault(position_and_feature_name, set()).add(
+                        feature_value
+                    )
+        # assign a unique index to each feature value
+        return self._build_feature_to_index_map(feature_vocabulary)
+
+    def _map_tokens_to_raw_features(
+        self, tokens: List[Token]
+    ) -> List[Dict[Tuple[int, Text], Hashable]]:
+        """Extracts the raw feature values.
+
+        Args:
+          tokens: a tokenized text
+        Returns:
+          a list of feature dictionaries for each token in the given list
+          where each feature dictionary maps a tuple containing
+          - a position (in the window) and
+          - a supported feature name
+          to the corresponding raw feature value
+        """
         sentence_features = []
 
-        for token_idx in range(len(tokens)):
-            # get the window size (e.g. before, word, after) of the configured features
+        for anchor in range(len(tokens)):
+
             # in case of an even number we will look at one more word before,
             # e.g. window size 4 will result in a window range of
             # [-2, -1, 0, 1] (0 = current word in sentence)
-            window_size = len(configured_features)
+            window_size = len(self._feature_config)
             half_window_size = window_size // 2
             window_range = range(-half_window_size, half_window_size + window_size % 2)
+            assert len(window_range) == window_size
```

Comment:
Can this happen outside of the loop? (not dep. on anchor)

Reply:
oh, good point! will move that